TerraForge3D/src/Generators/CPUNodeEditor/Nodes/HeightmapNode.cpp

```diff
@@ -0,0 +1,271 @@
+	}
+
+	if(npScale)
+	{
+		res = res * 2.0f - 1.0f;
+	}
+
+	if(inv)
+	{
+		res = 1.0f - res;
+	}
+
+	mutex.unlock();
+	return NodeOutput({ res });
+}
+
+void HeightmapNode::Load(nlohmann::json data)
+{
+	scale = data["scale"];
+
+	if (isDefault && data["isDefault"])
+	{
+		return;
+	}
+
+	isDefault = data["isDefault"];
+	interpolated = data["interpolated"];
+	npScale = data["npsc"];
+	inv = data["inv"];
+	autoTiled = data["autoTiled"];
+	numTiles = data["numTiles"];
+	posi[0] = data["posiX"];
+	posi[1] = data["posiY"];
+	rota = data["rota"];
+
+	if (isDefault)
+	{
+		delete heightmap;
+		heightmap = new Heightmap(GetExecutableDir() + PATH_SEPARATOR "Data" PATH_SEPARATOR "heightmaps" PATH_SEPARATOR "flat.png");
+	}
+
+	else
+	{
+		std::string hash = data["heightmap"];
+
+		if (!ProjectManager::Get()->AssetExists(hash))
+		{
+			ShowMessageBox("Failed to Load Heightmap : " + hash, "Error");
+			isDefault = true;
+		}
+
+		else
+		{
+			delete heightmap;
+			heightmap = new Heightmap(ProjectManager::Get()->GetResourcePath() + PATH_SEPARATOR + ProjectManager::Get()->GetAsset(hash));
+			Log("Loaded Cached Heightmap : " + hash);
+		}
+	}
+}
+
+nlohmann::json HeightmapNode::Save()
+{
+	nlohmann::json data;
+	data["type"] = MeshNodeEditor::MeshNodeType::Heightmap;
+	data["scale"] = scale;
+	data["isDefault"] = isDefault;
+	data["interpolated"] = interpolated;
+	data["inv"] = inv;
+	data["npsc"] = npScale;
+	data["autoTiled"] = autoTiled;
+	data["numTiles"] = numTiles;
+	data["posiX"] = posi[0];
+	data["posiY"] = posi[1];
+	data["rota"] = rota;
+
+	if (!isDefault)
+	{
+		std::string hash = MD5File(heightmap->GetPath()).ToString();
+		data["heightmap"] = hash;
+
+		if (!ProjectManager::Get()->AssetExists(hash))
+		{
+			ProjectManager::Get()->SaveResource("heightmaps", heightmap->GetPath());
+			Log("Cached " + heightmap->GetPath());
+		}
+
+		else
+		{
+			Log("Heightmap already cached : " + hash);
+		}
+	}
+
+	return data;
+}
+
+void HeightmapNode::OnRender()
+{
+	DrawHeader("Heightmap");
+	inputPins[0]->Render();
+	ImGui::Text("X");
+	ImGui::SameLine();
+	ImGui::Dummy(ImVec2(150, 10));
+	ImGui::SameLine(0.0f, ImGui::GetStyle().ItemInnerSpacing.x);
+	ImGui::Text("Height");
+	ImGui::SameLine(0.0f, ImGui::GetStyle().ItemInnerSpacing.x);
+	outputPins[0]->Render();
+	inputPins[1]->Render();
+	ImGui::Text("Y");
+	inputPins[2]->Render();
+
+	if (inputPins[2]->IsLinked())
+	{
+		ImGui::Text("Scale");
+	}
+
+	else
+	{
+		ImGui::DragFloat(("Scale##" + std::to_string(inputPins[1]->id)).c_str(), &scale, 0.01f);
+	}
+
+	ImGui::Checkbox(("Auto Tiled##tild" + std::to_string(id)).c_str(), &autoTiled);
+	ImGui::Checkbox(("Interpolated##intrp" + std::to_string(id)).c_str(), &interpolated);
+	ImGui::Checkbox(("Inverted##tinv" + std::to_string(id)).c_str(), &inv);
+	ImGui::Checkbox(("Scale -1 To 1##tnpsc" + std::to_string(id)).c_str(), &npScale);
+
+	if(!autoTiled)
+	{
+		ImGui::DragFloat(("Num Tiles##nmtl" + std::to_string(id)).c_str(), &numTiles, 0.01f);
```

Comment:
Call ImGui::PushItemWidth(100) before any drag float, dragdloatn and after that call ImGui::PopItemWidth() else the long inputs sometimes causes inconvinience while using the node editor